app/models/miq_provision_virt_workflow.rb

```diff
@@ -216,15 +216,19 @@ def available_vlans_and_hosts(options = {})
       rails_logger('allowed_vlans', 0)
       # TODO: Use Active Record to preload this data?
       MiqPreloader.preload(hosts, :switches => :lans)
-      hosts.each do |h|
-        h.lans.each { |l| vlans[l.name] = l.name unless l.switch.shared? }
-      end
+      load_allowed_vlans(hosts, vlans)
       rails_logger('allowed_vlans', 1)
     end
 
     return vlans, hosts
   end
 
+  def load_allowed_vlans(hosts, vlans)
+    hosts.each do |h|
+      h.lans.each { |l| vlans[l.name] = l.name if @vlan_options[:dvs] || !l.switch.shared? }
```

Comment:
could you pass @vlan_options as a parameter? 

def load_allowed_vlans(hosts, vlans, vlan_options = [])

@agrare ^ this should not break VMware? Not sure if we have specs for this particular case.